go/vt/vttablet/tabletmanager/tm_state.go

```diff
@@ -186,6 +186,9 @@ func (ts *tmState) ChangeTabletType(ctx context.Context, tabletType topodatapb.T
 			// We must read the data again and verify whether the previous write succeeded or not.
 			// The only way to guarantee safety is to keep retrying read until we succeed
 			for {
+				if ctx.Err() != nil {
+					return ctx.Err()
```

Comment:
We had added the infinite loop intentionally, because we enter that code only if a topo write has failed. Once a topo write fails, the only way forward is to keep trying to read the topo server, because the information may or may not have been written.
However, I do understand the use case where they want to retry doing the same operation again, so we need to exit sometime and not wait indefinitely.

I just think we should wrap the returned error and explicitly say that the topo information might be different from the tablet information and that the user must call ChangeTabletType again to fix it.
Other than that, this looks good to me

Reply:
Yes I think that will work!